pkg/controller/promotionreconciler/prowjobretrigger/reconciler.go

```diff
@@ -0,0 +1,185 @@
+	controllerutil "github.com/openshift/ci-tools/pkg/controller/util"
+	"github.com/openshift/ci-tools/pkg/load/agents"
+)
+
+type Options struct {
+	DryRun                bool
+	CIOperatorConfigAgent agents.ConfigAgent
+	ConfigGetter          config.Getter
+	GitHubClient          github.Client
+}
+
+const (
+	ControllerName = "promotion_job_retrigger"
+)
+
+func AddToManager(mgr controllerruntime.Manager, opts Options) error {
+	prowJobEnqueuer, err := prowjobreconciler.AddToManager(mgr, opts.ConfigGetter, opts.DryRun)
+	if err != nil {
+		return fmt.Errorf("failed to construct prowjobreconciler: %w", err)
+	}
+
+	log := logrus.WithField("controller", ControllerName)
+	r := &reconciler{
+		ctx:          context.Background(),
+		log:          log,
+		client:       mgr.GetClient(),
+		gitHubClient: opts.GitHubClient,
+		enqueueJob:   prowJobEnqueuer,
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler:              r,
+		MaxConcurrentReconciles: 10,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+	if err := c.Watch(
+		&source.Kind{Type: &prowjobsv1.ProwJob{}},
+		handler.Funcs{
+			UpdateFunc: func(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+				q.Add(e)
+			},
+		},
+		predicate.NewPredicateFuncs(func(meta metav1.Object, object runtime.Object) bool {
+			prowJob, ok := object.(*prowjobsv1.ProwJob)
+			if !ok {
+				return false
+			}
+			if !strings.HasSuffix(prowJob.Spec.Job, "-images") {
+				return false
+			}
+			return prowJob.Status.State == prowjobsv1.FailureState || prowJob.Status.State == prowjobsv1.ErrorState
+		}),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for ProwJobs: %w", err)
+	}
+	r.log.Info("Successfully added reconciler to manager")
+
+	return nil
+}
+
+type RefGetter interface {
+	GetRef(org, repo, ref string) (string, error)
+}
+
+type reconciler struct {
+	ctx          context.Context
+	log          *logrus.Entry
+	client       ctrlruntimeclient.Client
+	gitHubClient RefGetter
+	enqueueJob   prowjobreconciler.Enqueuer
+}
+
+func (r *reconciler) Reconcile(req controllerruntime.Request) (controllerruntime.Result, error) {
+	log := r.log.WithField("name", req.Name).WithField("namespace", req.Namespace)
+	log.Trace("Starting reconciliation")
+	startTime := time.Now()
+	defer func() { log.WithField("duration", time.Since(startTime)).Trace("Finished reconciliation") }()
+
+	err := r.reconcile(req, log)
+	if err != nil {
+		log := log.WithError(err)
+		// Degrade terminal errors to debug, they most lilely just mean a given imageStreamTag wasn't built
+		// via ci operator.
+		if controllerutil.IsTerminal(err) {
+			log.Debug("Reconciliation failed")
+		} else {
+			log.Error("Reconciliation failed")
+		}
+	}
+
+	return controllerruntime.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+func (r *reconciler) reconcile(req controllerruntime.Request, log *logrus.Entry) error {
```

Comment:
We need something ton ratelimit this in case of jobs that are broken

Reply:
I'm not sure we do - we don't enqueue new jobs if one is running and images postsubmits have `max_concurrency` set to 1 so we're effectively rate-limited at serial retry. Not sure how exactly we could achieve e.g. exponential backoff for this since it's not one specific key in the queue that fails, the queue keys get processed correctly and it's the ensuing job that fails. I will add alerts for this controller doing too many re-runs and we *should* be OK with the serial behavior?